### ISSUE TITLE
[bug][16.5.0] option returns [object Object] instead string

### ISSUE THREAD

**Issue by alxdrk** [facebook/react#13586]
**What is the current behavior?**
`<option>` returns [object Object] instead string

**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem. Your bug will get fixed much faster if we can run your code and it doesn't have dependencies other than React. Paste the link to your JSFiddle (https://jsfiddle.net/Luktwrdm/) or CodeSandbox (https://codesandbox.io/s/new) example below:**
Demo: https://codesandbox.io/s/ww5mv2w957

**What is the expected behavior?**
Expected string

**Which versions of React, and which browser / OS are affected by this issue? Did this work in previous versions of React?**

**Comment by jquense** [comment]
`FormattedMessage` doesn't render a string, it renders a `span` which is not valid HTML for the inside of an `option` tag. You should use `intl.formatMessage()` directly or try setting the https://github.com/yahoo/react-intl/wiki/Components#intlprovider `textcomponent` to `React.Fragment` to avoid the extra span

**Comment by alxdrk** [comment]
@jquense 
Sorry, forgot add default tag for `FormattedMessage`
It doesn't work with fragment too.

**Comment by bvaughn** [comment]
For what it's worth, the same Code Sandbox works with React 16.4 but not 16.5.

**Comment by bvaughn** [comment]
Maybe related to #13465?

**Comment by Simek** [comment]
If you remove `option` tag and add `tagName` property to `FormattedMessage` demo works correctly.

Fork: https://codesandbox.io/s/6yr4xy5ll3

**Comment by bvaughn** [comment]
cc @gaearon

**Comment by jquense** [comment]
> It doesn't work with React.Fragment too.

we should make this work probably as well.

IDK how to categorize this breakage, it worked solely because of how forgiving browsers can be  (which is not unlimited causing the original bug), but the markup is definitely incorrect. It also definitely broke something that "worked" before

**Comment by quazzie** [comment]
Does not work with a context consumer inside a option either.
https://codesandbox.io/s/pk574rm207

**Comment by bvaughn** [comment]
I'm going to tag this as a regression for now, at least until we're sure one way or another.

**Comment by Simek** [comment]
@quazzie You can fix that by wrapping `value` in `option` tag inside context `Consumer` render function.

Fork: https://codesandbox.io/s/oom712ov79

**Comment by quazzie** [comment]
@Simek thanks but in my real app i use a component that uses context internally so i can't really use that.
```<option><MyComponent value="value" /></option>```
and MyComponent uses a context.Consumer, processes value a bit and returns a string.

**Comment by quazzie** [comment]
@bvaughn yeah it seems to be connected to #13465
It seems you can't use any component as child to option.
Test with simple functional component yields same problem:
https://codesandbox.io/s/7y5km7r7k6

**Comment by Simek** [comment]
@quazzie Sure thing, it seems like it's a regression. 

For now if you want you can implement `tagName` property and pass `option` as value to component in your app. If property has been provided you can wrap String inside component otherwise you can just return String.

To be clear I'm just providing quick solutions until React team decided what to do with that issue.

**Comment by gaearon** [comment]
I don't think it was a regression. This kind of a thorny area. It was never intentionally supported. It accidentally worked on initial mount but then crashed on updates (https://github.com/facebook/react/pull/13261). Fixing the crash was more important, so we fixed it to be treated as text content (which it should be). Unfortunately this means putting custom components in the middle is not supported. That's consistent with how `textarea` and similar elements work.

I think it's better to show invalid output and warn about something that breaks on updates, than to let people use it only to discover it crashes in production. But I can see arguments for why this should be supported when the custom component returns a string.

**Comment by nicolas-cherel** [comment]
@gaearon a single `children` that renders a single string wouldn't trigger the crash. I understand that it's easier to validate and enforce `<option>` children input, but the truth is, anyone that want to use context to setup an `<option>` has to leak the context into the parent. That, and having a corner case where you can't use react to setup an element children feels extremely weird.

**Comment by gaearon** [comment]
@nicolas-cherel You're welcome to file a proposal for how it should work (as long as it doesn't include crashes).

**Comment by gaearon** [comment]
To be clear — if you figure out how to make it work in all cases, I think we'd be happy supporting using custom components returning strings for this. I couldn't figure it out yet.

**Comment by exogen** [comment]
Shouldn't this be kept open (or is there another issue tracking this?) even if nobody has any implementation ideas yet? It's totally reasonable to expect `<option><React.Fragment>Foo</React.Fragment></option>` to work, considering the developer knows that the output HTML would in fact be valid.

**Comment by gaearon** [comment]
As a feature request, sure. Please file a new issue or search if one exists?

**Comment by tobico** [comment]
@gaearon I see how this would be difficult to fix, but I'd argue that it still counts as a bug. Isn't it reasonable for a user of React to expect `option` elements to support nesting of components the same way that every other React component does? If it says anywhere in the React docs and tutorial that only certain kinds of components can be nested, I must have missed it.

**Comment by Jokinen** [comment]
>React 15 also warned about this but we accidentally lost the warning in 16 for cases when components return a string.

I am sorry to go a bit off track here, but what is the status for this lost warning? Is it something which won't be re-introduced?

We had the exact case where we were using a translation component returning a string. The end result is that `[object Object]` gets rendered which is quite unexpected in the context of React.

**Comment by vandercloak** [comment]
Just opened a feature request here https://github.com/facebook/react/issues/15513

**Comment by lukescott** [comment]
I find it rather surprising that this:

```jsx
<option><FormatMessage message={message /><option>
```
Returns:
```jsx
<option>[object Object]</option>
```
In a browser you can do `<option><span>foo</span></option>` and it does not care. Chrome strips the tags. So I'm not sure why React cares, especially since `<FormatMessage />` is returning a string (or a fragment that results in a string).

IMO, even if this was not working in React 15 it should still be considered a bug. In my case the work-around isn't as easy.

**Comment by lukescott** [comment]
@gaearon Maybe I'm missing something, but 16.4 did not warn about this code:

```js
import React from "react";
import ReactDOM from "react-dom";

const App = () => {
  return (
    <select>
      <option>
        <Foo />
      </option>
    </select>
  );
};

function Foo() {
  return "foo";
}

ReactDOM.render(<App />, document.getElementById("root"));
```

This code worked just fine in React 16.4, but later broke in 16.5. In some of the code examples I've seen from `react-intl` package, it would render as `<option><span>message</span></option>` which would cause the warning. But in the example above no such warning occurs because the component returns a string.

16.4.2: https://codesandbox.io/s/gifted-beaver-xjv1j
16.8.6: https://codesandbox.io/s/angry-carson-8rnd3

Could we perhaps reevaluate this as a regression that needs to be fixed rather than an enhancement?

I would expect that `<option><span>message</span></option>` should be an error while `<option>foo</option>` (from the example above) should not.

**Comment by Slatery** [comment]
If anyone still needs an answer to the original question, as mentioned in earlier comments and in the documentation [here,](https://github.com/formatjs/react-intl/blob/master/docs/Components.md#formattedmessage) you cannot use a formatted message in the format of 
```
<select>
    <option value="foo">
        <FormattedMessage id="bar"/>
    </option>
</select>
```
instead I've found formating it as such does work:
```
<select>
    <FormattedMessage id="bar">
        {txt => <option value="foo">{txt}</option>}
    </FormattedMessage>
</select>
```
where `txt` will return the id's default string for whichever Message Descriptor file is selected

**Comment by danielo515** [comment]
So any valid workaround that doesn't involves changing your entire component tree? Many people is using context inside options to provide localisation etc.

**Comment by hrupesh** [comment]
Is there any solution to this till date?

**Comment by wojtekmaj** [comment]
@hrupesh There's still no solution that isn't hacky. What I did and what my personal recommendation would be is to use a replacement over native select element, carefully crafted to keep accessibility levels high, like [Reach UI Listbox](https://reach.tech/listbox).

**Comment by gaearon** [comment]
This is fixed in 18 Alpha.
https://codesandbox.io/s/thirsty-austin-yzgyg?file=/src/index.js

**Comment by gaearon** [comment]
React 18 is released with the fix.
You need to pass `value` to `<option>` for this.

https://codesandbox.io/s/epic-sea-r219ps?file=/src/index.js
